systests/jaxrs/src/test/java/org/apache/cxf/systest/jaxrs/BookServer20.java

```diff
@@ -227,7 +227,7 @@ public void filter(ContainerRequestContext context) throws IOException {
             } else if (path.endsWith("books/check2")) {
                 replaceStream(context);
             } else if (path.endsWith("books/checkNQuery")) {
-                URI requestURI = URI.create(path.replace("NQuery", "2?a=b"));
+                URI requestURI = URI.create(path.replace("NQuery", "2"));
```

Comment:
I was debugging this failure and it seems like the change with respect to Request URI 

```
 HttpUtils.resetRequestURI(m, requestUri.toString());
```

violates some CXF flow (it does not mean the flow is correct, it just probably means we need changes in other places). 

In the nutshell, my findings are as such:
 - CXF assumes the `m.get(Message.REQUEST_URI)` has only URI raw path component
 - CXF assumes the `m.get(Message.QUERY_STRING)` has only URI query string component

The fact that we now set the full URI (with a query string) breaks the way CXF treats the  `m.get(Message.REQUEST_URI)` and test fails because as you correctly pointing out,  the query string is messed up `/bookstore/books/check2?a=b?a=b`. 

I think although your change does fix the test, the flow is broken by and large, please correct me if I am wrong. The user should be able to change the Request URI in pre-matching filter, including the query string component, but she won't be able to do that from now on.

Reply:
> I think although your change does fix the test, the flow is broken by and large, please correct me if I am wrong. The user should be able to change the Request URI in pre-matching filter, including the query string component, but she won't be able to do that from now on.

Hmm... maybe I don't fully understand.  I think the test was broken from the start since it was effectively setting the RequestUri to `/bookstore/books/check2?a=b?a=b` - the previous behavior (prior to #697) was to simply ignore the query string entirely which was why the test passed even though the URI was invalid.  After #697 the test failed because we were actually checking the query string.  

Do you think we need to explicitly set `Message.QUERY_STRING` when the user calls `containerRequestContext.setRequestUri(newUri)`? In that case then we might also need to do some checking and "URI merging"  in other places where JAX-RS related code is expecting to see the full URI (including the query string) - which is the point you were making, I think.

Do you know which places in the code are expecting `Message.REQUEST_URI` to be the raw URI?  I'd like to take a look to determine whether we should change those parts to expect REQUEST_URI to be the full URI or whether we should change the JAX-RS parts to merge the REQUEST_URI with the QUERY_STRING when checked.

Thanks for the review and the analysis.

Reply:
> Hmm... maybe I don't fully understand. I think the test was broken from the start since it was effectively setting the RequestUri to `/bookstore/books/check2?a=b?a=b` - the previous behavior (prior to #697) was to simply ignore the query string entirely which was why the test passed even though the URI was invalid. After #697 the test failed because we were actually checking the query string.

I think the test case was fine, let me try to illustrate what was happening before the change and after, for simplicity I will just include the URI into method arguments.

So before:

```
containerRequestContext.setRequestUri("/bookstore/books/check2?a=b")

m.get(Message.REQUEST_URI) -> /bookstore/books/check2
m.get(Message.QUERY_STRING) -> a=b
```

After the change:

```
containerRequestContext.setRequestUri("/bookstore/books/check2?a=b")

m.get(Message.REQUEST_URI) -> /bookstore/books/check2?a=b
m.get(Message.QUERY_STRING) -> a=b
```

As you can see, `Message.REQUEST_URI` has raw path before, but now - full URI. And that causing the issues. Unfortunately I don't know all the places but it seems like raw URI is being used to find the matching methods to invoke fe.

Reply:
Ah, that makes sense.  For now, I'm reverting that portion of the change in ContainerRequestContextImpl - I added a TODO so that we won't lose sight of it, but that should get the tests passing again.  Now I'll take a look at the OIDC tests that are failing.  It's up to you whether we should merge this now to get the JAX-RS systests back to 100% passing or wait until we get the OIDC tests passing too.  Thanks again!

Reply:
Thanks @andymc12 , now we have better understanding of the impact.